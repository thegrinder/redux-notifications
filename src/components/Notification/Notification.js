import styled from 'styled-components';
import PropTypes from 'prop-types';
import { convertToCssDuration, getAnimatedMargin } from '../../helpers/utils';
import { verticalShow, verticalHide } from '../../animations/animations';

const propTypes = {
  isVisible: PropTypes.bool.isRequired,
  notificationHeight: PropTypes.number.isRequired,
  position: PropTypes.array.isRequired,
  animationDuration: PropTypes.number.isRequired,
  animationEasing: PropTypes.string.isRequired,
};

const Notification = styled.div`
  position: relative;
  z-index: ${(props) => (props.isVisible ? 999999 : -1)};
  animation: ${({ isVisible, position, notificationHeight }) => {
    const animatedMargin = getAnimatedMargin(position);
    return isVisible
      ? verticalShow(notificationHeight, animatedMargin)
      : verticalHide(notificationHeight, animatedMargin);
  }};
  animation-timing-function: ${(props) => props.animationEasing};
  animation-duration: ${(props) =>
    convertToCssDuration(props.animationDuration)};
  animation-fill-mode: forwards;
`;

Notification.propTypes = propTypes;

export default Notification;
