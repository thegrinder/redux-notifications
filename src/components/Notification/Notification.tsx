import React, { useLayoutEffect, useRef, useState, ReactNode } from 'react';
import {
  convertToCssDuration,
  getAnimatedMargin,
  capitalize,
  Position,
} from '../../helpers/utils';

type NotificationProps = {
  position: Position;
  isVisible: boolean;
  animationDuration: number;
  animationEasing: string;
  children: ReactNode;
};

const Notification = ({
  position,
  isVisible,
  animationDuration,
  animationEasing,
  children,
}: NotificationProps) => {
  const notificationRef = useRef<HTMLDivElement>(null);
  const [height, updateHeight] = useState(0);

  useLayoutEffect(() => {
    if (notificationRef.current) {
      updateHeight(notificationRef.current.clientHeight);
    }
  }, [notificationRef]);

  return (
    <div
      ref={notificationRef}
      style={{
        position: 'relative',
        zIndex: isVisible ? 999999 : -1,
        transitionProperty: `margin-${getAnimatedMargin(position)}, opacity`,
        transitionDuration: convertToCssDuration(animationDuration),
        transitionTimingFunction: animationEasing,
        [`margin${capitalize(getAnimatedMargin(position))}`]: isVisible
          ? 0
          : `${-height}px`,
        opacity: isVisible ? 1 : 0,
      }}
    >
      {children}
    </div>
  );
};

export default Notification;
