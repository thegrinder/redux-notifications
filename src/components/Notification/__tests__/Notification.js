import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Notification from '../Notification';

const testProps = {
  isVisible: true,
  notificationHeight: 40,
  animationDuration: 400,
  animationEasing: 'ease',
  slideFromSide: 'left',
  position: ['10px', '10px', 'auto', 'auto'],
};
const children = <p>test</p>;
const renderComponent = (props = {}) =>
  render(<Notification {...props}>{children}</Notification>);

describe('<Notification />', () => {
  afterEach(cleanup);

  it('should render correctly if visible', () => {
    const {
      container: { firstChild },
    } = renderComponent(testProps);
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should render correctly if not visible', () => {
    const {
      container: { firstChild },
    } = renderComponent({ ...testProps, isVisible: false });
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should have children', () => {
    const {
      container: { firstChild },
      getByText,
    } = renderComponent(testProps);
    const childrenElement = getByText('test');
    expect(firstChild).toContainElement(childrenElement);
  });
});
