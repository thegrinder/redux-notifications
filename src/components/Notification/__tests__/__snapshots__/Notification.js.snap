// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> should render correctly if not visible 1`] = `
<div
  class="sc-AxjAm hYfowd"
>
  <p>
    test
  </p>
</div>
`;

exports[`<Notification /> should render correctly if visible 1`] = `
<div
  class="sc-AxjAm kfvNSr"
>
  <p>
    test
  </p>
</div>
`;
